gio-subclass/src/auto/action.rs

```diff
@@ -0,0 +1,238 @@
+
+}
+
+any_impl!(ActionImpl);
+
+pub trait ActionImplStatic<T: ObjectType>: 'static {
+    fn get_impl<'a>(&self, imp: &'a T::ImplType) -> &'a ActionImpl;
+}
+
+struct ActionStatic<T: ObjectType>{
+    imp_static: *const ActionImplStatic<T>
+}
+
+// FIXME: Boilerplate
+unsafe extern "C" fn action_activate<T: ObjectType>
+(gptr: *mut gio_ffi::GAction, parameter: *mut glib_ffi::GVariant)
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+
+    let param = if parameter.is_null(){
+        None
+    }else{
+        Some(from_glib_none(parameter))
+    };
+    imp.activate(&wrap, param.as_ref())
+}
+
+unsafe extern "C" fn action_change_state<T: ObjectType>
+(gptr: *mut gio_ffi::GAction, value: *mut glib_ffi::GVariant)
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.change_state(&wrap, &from_glib_none(value))
+}
+
+unsafe extern "C" fn action_get_enabled<T: ObjectType>
+(gptr: *mut gio_ffi::GAction) -> gboolean
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.get_enabled(&wrap).to_glib()
+}
+
+unsafe extern "C" fn action_get_name<T: ObjectType>
+(gptr: *mut gio_ffi::GAction) -> *const c_char
+{
+    callback_guard!();
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_get_type())
+                                     as *const ActionStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    match imp.get_name(&wrap){ Some(t)  => t/*Not checked*/.to_glib_none().0, None => ptr::null()}
```

Comment:
This is unfortunately unsafe (and `get_parameter_type()` and `get_state_type()`). The pointers returned need to be valid after returning (by using `to_glib_none` they stop being valid after the scope of the function is gone).

A possible solution for this would be doing it like the protocol stuff in `GstURIHandler`, see https://github.com/sdroege/gst-plugin-rs/blob/754373ba6c89cc2bb9fe5779b156dcf0e374a465/gst-plugin/src/uri_handler.rs#L32 (and wherever it is used). Here these "static" values are part of the `Static` part of the interface and it ensures that the underlying pointers are always valid.

However for your case you need them to be valid per instance, not per type (correct?). So a solution to that would be to require them to be given during instance initialization (by having some public function that has to be called then), and to store them externally in the instance so that the implementation can't modify them. E.g. via `g_object_get_qdata` and the `g_object_set_qdata_full`. That way you have control over the lifetime of the pointers instead of the implementation.

Reply:
Pfew. I'm not sure how I could get that to look clean though...

Reply:
Or you could do it all in the trampoline for the vfunc. You check if some qdata already exists, if so you just use that. Otherwise (i.g. the first time) you call the impl for the return value and then store it in the qdata.

Reply:
That seems like a better approach to me. Seems silly to store a return value somewhere in user-impls. As far as rust is concerned, the caller should no longer care about the value.

Reply:
This could in theory also be autogenerated from the code generator, but better just do it manually for the cases for now and let's see if we find a weird case.